File: packages/journal/Tee.py
# -*- coding: utf-8 -*-
#
# michael a.g. aïvázis <michael.aivazis@para-sim.com>
# (c) 1998-2026 all rights reserved


# superclass
from .Splitter import Splitter

# the devices i assemble
from .Console import Console
from .File import File


# a splitter that writes every entry to the console and to a set of files
class Tee(Splitter):
    """
    Journal device that writes every entry to the console and to a file at each of a set of
    paths
    """

    # constants
    name = "tee"

    # metamethods
    def __init__(self, paths=(), name=name, **kwds):
        # the console comes first, then a file at each path, open for writing
        outputs = [Console()] + [File(path=path) for path in paths]
        # chain up
        super().__init__(outputs=outputs, name=name, **kwds)
        # all done
        return


# end of file

File: lib/journal/Tee.cc
// -*- C++ -*-
// -*- coding: utf-8 -*-
//
// michael a.g. aïvázis <michael.aivazis@para-sim.com>
// (c) 1998-2026 all rights reserved


// external support
#include "externals.h"
// forward declarations
#include "forward.h"
// type aliases
#include "api.h"

// get the support i need
#include "Entry.h"
#include "Device.h"
#include "Splitter.h"
#include "Console.h"
#include "File.h"
// my declarations
#include "Tee.h"


// metamethods
// a tee to the console and to a file at each of the given {paths}
pyre::journal::Tee::Tee(const paths_type & paths, const name_type & name) : super_type(name)
{
    // the console comes first
    attach(std::make_shared<Console>());
    // then a file at each path
    for (const auto & path : paths) {
        // that opens for writing
        attach(std::make_shared<File>(path));
    }
    // all done
    return;
}


// destructor
pyre::journal::Tee::~Tee() {}


// end of file
